cloud/ec2_fleet.go

```diff
@@ -0,0 +1,538 @@
+	if err != nil {
+		return false, errors.Wrap(err, "error checking if instance is up")
+	}
+	if status == StatusRunning {
+		return true, nil
+	}
+	return false, nil
+}
+
+// OnUp is a noop for ec2Fleet
+func (m *ec2FleetManager) OnUp(context.Context, *host.Host) error {
+	return nil
+}
+
+func (m *ec2FleetManager) GetDNSName(ctx context.Context, h *host.Host) (string, error) {
+	r, err := getRegion(h)
+	if err != nil {
+		return "", errors.Wrap(err, "problem getting region from host")
+	}
+	if err = m.client.Create(m.credentials, r); err != nil {
+		return "", errors.Wrap(err, "error creating client")
+	}
+	defer m.client.Close()
+
+	instance, err := m.getInstance(ctx, h)
+	if err != nil {
+		return "", errors.Wrapf(err, "can't get instance information for '%s", h.Id)
+	}
+
+	// Cache availability zone, launch time, and volume size in host document, since we
+	// have access to this information now. Cost jobs will use this
+	// information later.
+	h.Zone = *instance.Placement.AvailabilityZone
+	h.StartTime = *instance.LaunchTime
+	h.VolumeTotalSize, err = getVolumeSize(ctx, m.client, h)
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting volume size for host %s", h.Id)
+	}
+	if err = h.CacheHostData(); err != nil {
+		return "", errors.Wrap(err, "error updating host document in db")
+	}
+
+	// set IPv6 address, if applicable
+	for _, networkInterface := range instance.NetworkInterfaces {
+		if len(networkInterface.Ipv6Addresses) > 0 {
+			err = h.SetIPv6Address(*networkInterface.Ipv6Addresses[0].Ipv6Address)
+			if err != nil {
+				return "", errors.Wrap(err, "error setting ipv6 address")
+			}
+			break
+		}
+	}
+	return *instance.PublicDnsName, nil
+}
+
+func (m *ec2FleetManager) GetSSHOptions(h *host.Host, keyName string) ([]string, error) {
+	return getEc2SSHOptions(h, keyName)
+}
+
+func (m *ec2FleetManager) TimeTilNextPayment(h *host.Host) time.Duration {
+	return timeTilNextEC2Payment(h)
+}
+
+func (m *ec2FleetManager) getInstance(ctx context.Context, h *host.Host) (*ec2.Instance, error) {
+	instance, err := m.client.GetInstanceInfo(ctx, h.Id)
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message":       "error getting instance info",
+			"host":          h.Id,
+			"host_provider": h.Distro.Provider,
+			"distro":        h.Distro.Id,
+		}))
+		return nil, errors.Wrap(err, "error getting instance info")
+	}
+
+	if err = validateEc2InstanceInfoResponse(instance); err != nil {
+		return nil, errors.Wrap(err, "invalid instance info response")
+	}
+
+	return instance, nil
+}
+
+func (m *ec2FleetManager) spawnFleetSpotHost(ctx context.Context, h *host.Host, ec2Settings *EC2ProviderSettings, blockDevices []*ec2.LaunchTemplateBlockDeviceMappingRequest) ([]*string, error) {
+	// Cleanup
+	var templateID *string
+	defer func() {
+		if templateID != nil {
+			// Delete launch template
+			_, err := m.client.DeleteLaunchTemplate(ctx, &ec2.DeleteLaunchTemplateInput{LaunchTemplateId: templateID})
```

Comment:
For future work, would it make sense to keep an integer in the distro document, and increment it each time the distro is updated? Then, we could keep the launch templates around and not have to delete them if they haven't been modified.

Reply:
I think there might be added complexity with maintaining a collection of launch templates on AWS.
On the other hand, it would save API calls.